Source/Methods/Envelope Tracking/EnvelopeTracking.cs

```diff
@@ -0,0 +1,309 @@
+﻿using NationalInstruments.DataInfrastructure;
+using NationalInstruments.ModularInstruments;
+using NationalInstruments.ModularInstruments.NIRfsg;
+using NationalInstruments.ModularInstruments.SystemServices.TimingServices;
+using System;
+using System.Linq;
+
+namespace NationalInstruments.ReferenceDesignLibraries.Methods
+{
+    public static class EnvelopeTracking
+    {
+
+        #region Type Definitions
+        public struct EnvelopeGeneratorConfiguration
+        {
+            public RfsgTerminalConfiguration TerminalConfiguration;
+
+            public static EnvelopeGeneratorConfiguration GetDefault()
+            {
+                return new EnvelopeGeneratorConfiguration()
+                {
+                    TerminalConfiguration = RfsgTerminalConfiguration.Differential,
+                };
+            }
+        }
+
+        public struct TrackerConfiguration
+        {
+            public double InputImpedance_Ohms;
+            public double CommonModeOffset_V;
+            public double Gain_VperV;
+            public double OutputOffset_V;
+
+            public static TrackerConfiguration GetDefault()
+            {
+                return new TrackerConfiguration()
+                {
+                    InputImpedance_Ohms = 1e6,
+                    CommonModeOffset_V = 1.0,
+                    Gain_VperV = 2.5,
+                    OutputOffset_V = 2.55
+                };
+            }
+        }
+
+        public enum DetroughType { 
+            Exponential, 
+            Cosine, 
+            Power
+        };
+
+        public struct DetroughConfiguration
+        {
+            public DetroughType Type;
+            public double MinimumVoltage;
+            public double MaximumVoltage;
+            public double Exponent;
+
+            public static DetroughConfiguration GetDefault()
+            {
+                return new DetroughConfiguration()
+                {
+                    Type = DetroughType.Exponential,
+                    MinimumVoltage = 1.5,
+                    MaximumVoltage = 3.5,
+                    Exponent = 1.2
+                };
+            }
+        }
+
+        public struct LookUpTable
+        {
+            public float[] DutInputPower_dBm;
+            public float[] SupplyVoltage_V;
+        }
+
+        public struct SynchronizationConfiguration
+        {
+            public double RFDelayRange_s;
+            public double RFDelay_s;
+
+            public static SynchronizationConfiguration GetDefault()
+            {
+                return new SynchronizationConfiguration()
+                {
+                    RFDelayRange_s = 2e-6, // +/- 1us
+                    RFDelay_s = 0.0
+                };
+            }
+        }
+        #endregion
+
+        #region Envelope Creation
+        public static Waveform CreateEnvelopeWaveform(Waveform referenceWaveform, DetroughConfiguration detroughConfig)
+        {

```

Comment:
Create makes sense, but for consistency, should we call it "ConfigureEnvelopeWaveform"

Reply:
That's a good point.

Reply:
will change it

Reply:
resolved in db4d450c736cd71e94237e354bdd12423706dbe3